File: javascript/EPI/wellFormedString.js


const wellFormedString = (s) =>{
    const pairs = {
        '(':')',
        '[':']',
        '{':'}',
    }

    const stack = []
    for(const letter of s){
        if(letter in pairs)
            stack.push(letter)
        else{
            if(pairs[stack.pop()] != letter)
                return false
        }
    }
    return !pairs
}

console.log(wellFormedString('}'))

File: python/EPI/wellFormedString.py



def wellFormedString(s):
    stack = []
    d = {
        '(': ')',
        '[': ']',
        '{': '}',
    }
    for letter in s:
        if(letter in d):
            stack.append(letter)
        elif not stack or letter != d[stack.pop()]:
            return False
    return not stack

print(wellFormedString('(())'))
print(wellFormedString('(()'))
print(wellFormedString("({})[]"))